presto-hive-common/src/main/java/com/facebook/presto/hive/zorder/ZOrderCurveSearcher.java

```diff
@@ -0,0 +1,233 @@
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.lang.String.format;
+
+/**
+ * The ZOrderCurveSearcher class contains functions to search for addresses within given value ranges.
+ */
+public class ZOrderCurveSearcher
+{
+    public enum ReturnType {
+        INTEGER, LONG
+    }
+
+    private final ZOrder zOrder;
+
+    private final List<Integer> encodingBits;
+    private final int totalBitLength;
+    private final int maxBitLength;
+
+    private List<ZValueRange> ranges;
+
+    private final ArrayList<ZAddressRange> addressRanges;
+
+    private ReturnType returnType;
+
+    private static int[] dimensions;
+
+    private ZOrderCurveSearcher(ZOrder zOrder, List<Integer> encodingBits, int totalBitLength, int maxBitLength)
+    {
+        this.zOrder = zOrder;
+
+        this.encodingBits = encodingBits;
+        this.totalBitLength = totalBitLength;
+        this.maxBitLength = maxBitLength;
+
+        this.addressRanges = new ArrayList<>();
+    }
+
+    /**
+     * Public class "constructor" specifying the ZOrder class and initializing <code>dimensions</code>.
+     * <p/>
+     * The <code>dimensions</code> array stores how many values are stored at each level of the Z-curve.
+     */
+    public static ZOrderCurveSearcher createZOrderCurveSearcher(ZOrder zOrder, List<Integer> encodingBits, int totalBitLength, int maxBitLength)
+    {
+        initializeDimensions(encodingBits, maxBitLength);
+        return new ZOrderCurveSearcher(zOrder, encodingBits, totalBitLength, maxBitLength);
+    }
+
+    private static void initializeDimensions(List<Integer> encodingBits, int maxBitLength)
+    {
+        dimensions = new int[maxBitLength];
+        List<Integer> bitLengths = new ArrayList<>(encodingBits);
+        for (int dimensionIndex = maxBitLength - 1; dimensionIndex >= 0; dimensionIndex--) {
+            for (int bitLengthIndex = 0; bitLengthIndex < bitLengths.size(); bitLengthIndex++) {
+                if (bitLengths.get(bitLengthIndex) > 0) {
+                    dimensions[dimensionIndex]++;
+                    bitLengths.set(bitLengthIndex, bitLengths.get(bitLengthIndex) - 1);
+                }
+            }
+        }
+    }
+
+    void setReturnType(ReturnType returnType)
+    {
+        this.returnType = returnType;
+    }
+
+    /**
+     * Searches for and outputs ranges of addresses within certain ranges in each dimension.
+     * <p/>
+     * Every two elements in the output list represent the minimum and maximum of an address range.
+     * <p/>
+     * For example, if the list is {1, 5, 10, 14}, the address ranges are from {1, 5} and {10, 14}.
+     *
+     * @param ranges the list of minimum and maximum values of each dimension
+     * @return the list of address ranges within the given value ranges
+     */
+    List<ZAddressRange> zOrderCondensedSearchCurve(List<ZValueRange> ranges)
+    {
+        this.ranges = fillInNulls(ranges);
+
+        addressRanges.clear();
```

Comment:
Will a searcher be used multiple times? If we need to clear `addressRanges` every time, then maybe we should not make this a class field. Instead, a local variable might be a better fit.